File: thinking-categorically/domain-and-codomain/01-composition-requirements/composition.cs
// Func<TInput, TOutput> shows domain and codomain clearly
Func<string, int> parseNumber = input => int.Parse(input);
//   ^       ^
//   |       codomain (int)
//   domain (string)

Func<int, string> formatResult = value => $"Result: {value}";
//   ^    ^
//   |    codomain (string)
//   domain (int)

// Composition using method chaining
Func<string, string> processInput = input =>
    formatResult(parseNumber(input));
//   ^       ^
//   |       codomain (string)
//   domain (string)

// Example usage
Console.WriteLine(processInput("42")); // "Result: 42"

File: thinking-categorically/domain-and-codomain/01-composition-requirements/composition.ts
// Function type annotations show domain and codomain
const parseNumber: (input: string) => number = (input) => parseInt(input);
//                 ^               ^
//                 |               codomain (number)
//                 domain (string)

const formatResult: (value: number) => string = (value) => `Result: ${value}`;
//                  ^               ^
//                  |               codomain (string)
//                  domain (number)

// Function composition
const processInput: (input: string) => string = (input) =>
  formatResult(parseNumber(input));
//            ^          ^
//            |          domain matches codomain of parseNumber
//            codomain of parseNumber matches domain of formatResult

// Example usage
console.log(processInput("42")); // "Result: 42"